### ISSUE TITLE
Wrong (x, y) coordinates on generated PDF

### ISSUE THREAD

**Issue by A-Hassan-M** [wkhtmltopdf/wkhtmltopdf#4742]
The y coordinates is reversed in the generated pdf, when I try to use a library to extract or add image to pdf it's inserted flipped on the x axis.
Note: I am using PDFBox to insert image to pdf.

**Comment by PhilterPaper** [comment]
The "generated PDF" was produced by wkHTMLtoPDF? I've noticed that some systems (e.g., Chrome's output to PDF) flip and scale the coordinate system (0,0 at top left, A4 page height is 3300 units, etc.). Maybe PDFBox is expecting such a flipped/scaled coordinate system? Check the documentation; maybe there's a flag to put it back into regular coordinates. Or possibly vice-versa; I don't know if wkHTMLtoPDF is the offending party (non-standard coordinate system) here.

**Comment by A-Hassan-M** [comment]
I tried other PDFs which work correctly but with PDF generated from wkHtmlToPdf the coordinate and scale is totally different and cannot achieve what I require

**Comment by PhilterPaper** [comment]
Looking at a PDF produced by wkHTMLtoPDF, it looks like wkHTMLtoPDF is the guilty party. At the beginning of the stream displaying a page, I see `0.750000000 0 0 -0.750000000 28.5000000 813.500000 cm`. That's going to flip and scale the coordinate system, in the same manner as Chrome does. It's left in that mode at the end of the object (not reset). If another object is inserted after this one, it will start out with this odd coordinate system.

I've seen this before in PDF::Builder (Perl library for creating PDFs), where someone created a PDF via Chrome browser, and used PDF::Builder to add some annotation to it. Unless you _first_ add your own `cm` command to cancel out the odd coordinate system (on each page), the text comes out flipped, tiny, and in an odd place. You may need to do something similar instead of using PDFBox.

**Comment by ashkulz** [comment]
I'm not sure this can be fixed, sorry -- it's probably an artifact of the way Qt's PDF rendering works. Maybe use a reset like @PhilterPaper mentions?

**Comment by A-Hassan-M** [comment]
Yeah that's what I tried to do and it worked but it was an ugly workaround, I used PDFBox contents transform to flip and scale coordinate system I don't know if there is another way to reset things but if you know a better way please tell me.

**Comment by PhilterPaper** [comment]
With Chrome-produced PDFs (perhaps just when annotated with page header and footer), what I ended up doing was inserting my own `cm` instruction to cancel the odd coordinate system, before doing any further work (adding text, etc) on each page. It's possible to do that with a library such as PDF::Builder, and from what you say, it might be possible in PDFBox "contents transform" to do something similar. That's just the "ugly workaround" you'll need to do (at least, it should be the same corrective code per page). Maybe something can be inserted into wkHTMLtoPDF's C++ code to clean up (cancel) the coordinate system at the end of an object (or just wrap each page in `q...Q`).
